# Лабораторная работа №3.

Выполнил студент гр. 9383 Моисейченко Кирилл. Вариант 10.

## Цель работы:

Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

## Постановка задачи:

1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

## Выполнение работы:

### 1. Решение диофантова уравнения

$2451x-4142y=-76$

In [1]:
from sympy.solvers.diophantine.diophantine import diop_solve
from sympy import symbols

In [2]:
x, y = symbols("x, y", integer=True)
print("Множество всех решений:", diop_solve(2451*x-4142*y+76))

Множество всех решений: (218*t_0 + 196, 129*t_0 + 116)


Частное решение: $(196, 116)$

### 2. RSA-дешифрование

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 11,m = 34).
Шифрованное сообщение имеет вид    (17; 20; 30; 16). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “ё” и “ъ”).

In [3]:
from sympy.crypto.crypto import decipher_rsa
from sympy.crypto.crypto import rsa_private_key

m = 34
e = 11

p = 2
q = 17

In [4]:
private_key = rsa_private_key(p, q, e)

encrypted_word = [17, 20, 30, 16]
word = []
for value in encrypted_word:
    word.append(decipher_rsa(value, private_key))
print(word)

[17, 10, 4, 16]


Зашифрованное слово: ПИГО

### 3. Запись в формате ДНФ и КНФ булевой функции, заданной аналитическим выражением.
$f(x, y, z) = ((z xor y) (zx)) xor (y \vee (z \vee x))$

In [5]:
from sympy.logic.boolalg import to_cnf, to_dnf, Xor
X, Y, Z = symbols("X, Y, Z", integer=True)

In [8]:
f = (Xor((Xor(Z, Y) & (Z & X)), (Y | (Z | X))))

#### Конъюктивная нормальная форма:

In [9]:
to_cnf(f, simplify = True)

(X | Y | Z) & (Y | ~X | ~Z)

#### Дизъюнктивная нормальная форма:

In [10]:
to_dnf(f, simplify = True)

Y | (X & ~Z) | (Z & ~X)

## Вывод:

Практически применена библиотека символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.